### PR TITLE
Reindex Basic implementation - Structure Code

### PR DESCRIPTION
**Pull request by pengchen0692**
## Description
Implement Reindex Operation. 
As a basic implemention, it allows reindex operation be triggered and execute until completion. 

## Related issues
Addresses [AB#83018](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/83018)

## Testing
All tests pass

### CODE REVIEW

File: src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Orchestration.cs [hunk 1]
```diff
@@ -13,34 +13,38 @@
...
-        ///  The orchestration function to add extended query tags.
+        ///  The orchestration function to reindex tags.
         /// </summary>
         /// <param name="context">The context.</param>
         /// <param name="logger">The logger.</param>
         /// <returns>The task.</returns>
-        [FunctionName(nameof(AddAndReindexTagsAsync))]
-        public async Task AddAndReindexTagsAsync(
+        [FunctionName(nameof(ReindexTagsAsync))]
+        public async Task ReindexTagsAsync(
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Orchestration.cs:30]: I've done some additional brainstorming, and I think we can reduce this to 1 orchestration. But for now we can keep it separate as we have more discussions offline lol

File: src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/AddInstanceTableValuedParametersBuilder.cs [hunk 2]
```diff
@@ -25,17 +25,52 @@ internal static class AddInstanceTableValuedParametersBuilder
...
+
+            return Build(instance, queryTags,
+                (stringRow, longRows, doubleRows, dateTimeRows, personNameRows)
+                => new V2.AddInstanceTableValuedParameters(stringRow, longRows, doubleRows, dateTimeRows, personNameRows));
+        }
+
+        private static T Build<T>(
+            DicomDataset instance,
+            IEnumerable<QueryTag> queryTags,
+            Func<IEnumerable<InsertStringExtendedQueryTagTableTypeV1Row>,
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/AddInstanceTableValuedParametersBuilder.cs:64]: Should we make a `private` delegate for this?

File: src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/AddInstanceTableValuedParametersBuilder.cs [hunk 3]
```diff
@@ -25,17 +25,52 @@ internal static class AddInstanceTableValuedParametersBuilder
...
+        /// <param name="instance">The dicom dataset.</param>
+        /// <param name="queryTags">The index tags.</param>
+        public static V2.AddInstanceTableValuedParameters BuildV2(
+            DicomDataset instance,
+            IEnumerable<QueryTag> queryTags)
+        {
+            EnsureArg.IsNotNull(instance, nameof(instance));
+            EnsureArg.IsNotNull(queryTags, nameof(queryTags));
+
+            return Build(instance, queryTags,
```
Comment (wsugarman) [inline src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/AddInstanceTableValuedParametersBuilder.cs:56]: nit: place each argument on its own line

**Review by wsugarman — APPROVED** [review]

File: src/Microsoft.Health.Dicom.Core/Features/Store/WatermarkRange.cs [hunk 4]
```diff
@@ -3,24 +3,27 @@
 // Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
 // -------------------------------------------------------------------------------------------------
 
-using System.Collections.Generic;
-
-namespace Microsoft.Health.Dicom.Functions.Indexing.Models
+namespace Microsoft.Health.Dicom.Core.Features.Store
```
[the remark this answers is not in the document]
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Core/Features/Store/WatermarkRange.cs:6]: Moved to Model

File: src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Trigger.cs [hunk 5]
```diff
@@ -24,21 +23,20 @@ public partial class ReindexDurableFunction
...
         {
             EnsureArg.IsNotNull(request, nameof(request));
             EnsureArg.IsNotNull(client, nameof(client));
             EnsureArg.IsNotNull(logger, nameof(logger));
-            var extendedQueryTags = await request.Content.ReadAsAsync<List<AddExtendedQueryTagEntry>>();
-            logger.LogInformation("Start adding extended query tags {input}", extendedQueryTags);
-            string instanceId = await client.StartNewAsync(nameof(AddAndReindexTagsAsync), instanceId: null, extendedQueryTags);
+            var extendedQueryTagKeys = await request.Content.ReadAsAsync<IReadOnlyCollection<int>>();
+            logger.LogInformation("Start reindexing extended query tags {extendedQueryTagKeys}", extendedQueryTagKeys);
+            string instanceId = await client.StartNewAsync(nameof(ReindexTagsAsync), instanceId: null, extendedQueryTagKeys);
```
[the remark this answers is not in the document]
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Trigger.cs:37]: I'm not sure how to properly handle exception in this trigger. I'll leave a TODO here.

File: src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Orchestration.cs [hunk 6]
```diff
@@ -13,34 +13,38 @@
...
-            var input = context.GetInput<IReadOnlyList<AddExtendedQueryTagEntry>>();
-            var storeEntries = await context.CallActivityAsync<IReadOnlyList<ExtendedQueryTagStoreEntry>>(
-                nameof(AddTagsAsync),
-                input);
-            ReindexOperation reindexOperation = await context.CallActivityAsync<ReindexOperation>(
-                nameof(PrepareReindexingTagsAsync),
-                new PrepareReindexingTagsInput { OperationId = context.InstanceId, TagKeys = storeEntries.Select(x => x.Key).ToList() });
-            await context.CallSubOrchestratorAsync(nameof(ReindexTagsAsync), input: reindexOperation);
+            var input = context.GetInput<IReadOnlyCollection<int>>();
+            await context.CallActivityAsync(nameof(UpdateSchemaVersionAsync), null);
```
[the remark this answers is not in the document]
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.Functions/Indexing/ReindexDurableFunction.Orchestration.cs:37]: removed this activity

File: src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/AddInstanceTableValuedParametersBuilder.cs (hunk 2 above)
Comment (pengchen0692) [reply src/Microsoft.Health.Dicom.SqlServer/Features/ExtendedQueryTag/AddInstanceTableValuedParametersBuilder.cs:64]: yes

**Pull request merged**